pageserver/src/tenant/timeline.rs

```diff
@@ -3270,10 +3270,39 @@ impl Timeline {
                     let mut layers = self_clone.layers.write().unwrap();
                     let mut updates = layers.batch_update();
                     {
+                        use crate::tenant::layer_map::Replacement;
                         let l: Arc<dyn PersistentLayer> = remote_layer.clone();
-                        updates.remove_historic(l);
+                        match updates.replace_historic(&l, new_layer) {
+                            Ok(Replacement::Replaced { .. }) => { /* expected */ }
+                            Ok(Replacement::NotFound) => {
+                                // TODO: the downloaded file should probably be removed, otherwise
+                                // it will be added to the layermap on next load? we should
+                                // probably restart any get_reconstruct_data search as well.
+                                warn!("replacing downloaded layer into layermap failed because layer was not found");
+                            }
+                            Ok(Replacement::RemovalBuffered) => {
+                                unreachable!("current implementation does not remove anything")
+                            }
+                            Ok(Replacement::Unexpected(other)) => {
+                                // if the other layer would have the same pointer value as
+                                // expected, it means they differ only on vtables.
+                                //
+                                // otherwise there's no known reason for this to happen as
+                                // compacted layers should have different covering rectangle
+                                // leading to produce Replacement::NotFound.
+                                error!(
+                                    expected = ?Arc::as_ptr(&l),
+                                    other = ?Arc::as_ptr(&other),
```

Comment:
Will the addresses be of any use to us when looking at the logs?
I think type info and layer file name would be more useful, as they could spark some initial investigation.

Reply:
Good idea! Though I don't think there's any reason Layer's don't always impl Debug as well.

Reply:
Well, except for having a lot of fields I see.

Reply:
If it would have a different layer file name, it would also have a different `LayerKey` so it would not be found.

We could add `fn type_name() -> &'static str` to the trait, but I think that's what we should use `Debug` for... `short_id` is one way, but I don't think it could reveal anything we don't already have in the log, also it would be the same for Remote and Non-Remote.

Well, trying to figure something out.

Reply:
There is already a:

```rust
impl std::fmt::Debug for dyn Layer {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Layer")
            .field("short_id", &self.short_id())
            .finish()
    }
}
```

Not sure I see what is the point of this... I'll get rid of it and include a full derived Debug, it'll repeat a lot of already known context but I cannot see a way around it.

Reply:
I'll do this Debug impls in a follow-up, I hate when every PR has to touch every file in the repo. Well, even 5 more.